### PR TITLE
Allow custom beacon colors

### PR DESCRIPTION
**Pull request by gudenau**
Changed TileEntityBeacon to check the block for a custom beam colour to blend
Changed Block to add method to get custom beam colour

### CODE REVIEW

**Comment by diesieben07** [comment]
Everything in forge and vanilla uses the American English spelling "color", I suggest you change yours.

**Comment by gudenau** [comment]
Will do, how can I update the PR? New to git.

**Comment by covers1624** [comment]
Just pusha commit and it updates.

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch [hunk 1]
```diff
@@ -9,7 +9,27 @@
...
++                        tileentitybeacon$beamsegment.func_177262_a();
++                        continue;
++                    }
++                    afloat = colour;
++                }else {
++                    afloat = EntitySheep.func_175513_a((EnumDyeColor) iblockstate.func_177229_b(BlockStainedGlassPane.field_176245_a));
+                 }
+-
+-                afloat = EntitySheep.func_175513_a((EnumDyeColor)iblockstate.func_177229_b(BlockStainedGlassPane.field_176245_a));
+             }
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: patch kinda looks wonky here, why is it touching this line at all?

Comment (gudenau) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: It is because if I do not do that the color gets set to the value in the blockstate in the BlockStainedGlassPane, which will crash because it does not exist in most blocks.

**Comment by gudenau** [comment]
Updated code to change "colour" to "color".

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch (hunk 1 above)
Comment (HenryLoenwind) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: He meant, why do you edit the line? There is no change to anything but whitespace. You should leave it alone so your patch would only add liness before it and a line with a single "}" after it.

Same for the two lines above. Don't change the indent.

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch [hunk 2]
```diff
@@ -9,7 +9,27 @@
...
-@@ -181,7 +181,7 @@
+                         break;
+                     }
+ 
+-                    tileentitybeacon$beamsegment.func_177262_a();
+-                    continue;
++                    // Allow any block to change the beacon beam color
++                    float[] color = iblockstate.func_177230_c().getBeaconColor(field_145850_b, iblockstate);
++                    if(color == null || color.length != 3){
++                        tileentitybeacon$beamsegment.func_177262_a();
```
Comment (HenryLoenwind) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: I personally don't like the length check. It is a pain to debug things if something you code just does nothing. It is much simpler if it explodes in your face and you get a nice stack trace.

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch (hunk 1 above)
Comment (gudenau) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: Oh, I didn't mean to. Does that effect anything?

File: patches/minecraft/net/minecraft/block/Block.java.patch [hunk 3]
```diff
@@ -1354,12 +1354,25 @@
...
++     * Gets the color that the beacon beam should be blended with.
++     *
++     * This should return null for no change, and a three entry float array with RGB values from 0.0 to 1.0.
++     *
++     * @param world The world the block is in
++     * @param blockState The state of the block
++     * @return The float array of colors to blend
++     * */
++    public float[] getBeaconColor(World world, IBlockState blockState) {
++        return null;
```
Comment (HenryLoenwind) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: On the one hand that float[] is exactly what mc then needs to process the color, on the other hand it provides very little type safety. I'm a bit torn here, should I recommend you use a better return value or not?

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch (hunk 2 above)
Comment (gudenau) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: OK, should I throw an argument exception?

File: patches/minecraft/net/minecraft/block/Block.java.patch (hunk 3 above)
Comment (gudenau) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: I did have a length and null check, should I do bounds checking?

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch (hunk 1 above)
Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: We just want to keep patches small

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch (hunk 2 above)
Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: just use it directly (no checks), the runtime will do it for you :P 
It's a bit meh but we want to keep patch small

File: patches/minecraft/net/minecraft/block/Block.java.patch (hunk 3 above)
Comment (williewillus) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: I say just leave it as float[]

Comment (williewillus) [inline patches/minecraft/net/minecraft/block/Block.java.patch]: making it something else means extra patches to convert the data

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch [hunk 4]
```diff
@@ -9,7 +9,27 @@
                      {
                          this.field_146015_k = false;
                          this.field_174909_f.clear();
-@@ -181,7 +181,7 @@
+                         break;
+                     }
+ 
+-                    tileentitybeacon$beamsegment.func_177262_a();
+-                    continue;
++                    // Allow any block to change the beacon beam color
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: comment unneeded

File: patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch [hunk 5]
```diff
@@ -9,7 +9,27 @@
...
++                    float[] color = iblockstate.func_177230_c().getBeaconColor(field_145850_b, iblockstate);
++                    if(color == null){
++                        tileentitybeacon$beamsegment.func_177262_a();
++                        continue;
++                    }
```
Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: stuff the whole if statement onto one line to reduce patch size

Comment (HenryLoenwind) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: wrong direction, that part inside the {} is unchanged vanilla code that should not be part of the patch

Comment (gudenau) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: What do you mean by that?

Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: oh yes, that's bad. your patch takes vanilla code and moves it into that if statement. don't move the vanilla code, structure the patch around it

Comment (williewillus) [inline patches/minecraft/net/minecraft/tileentity/TileEntityBeacon.java.patch]: e.g. 
```
+ float[] color = iblockstate.func_177230_c().getBeaconColor(field_145850_b, iblockstate);
+ if(color==null){
OLD VANILLA CODE HERE
+ }
```

**Comment by williewillus** [comment]
Also, please don't use the Github editor to edit your patches, do it in an IDE and regenerate them properly using gradle

**Comment by gudenau** [comment]
Guess I'll just rebuild the thing and be more carefull, give me a few.

**Comment by gudenau** [comment]
Ok, I goofed this up pretty badly it seems.

Could someone else get this in correctly?
